test/integration/scheduler/framework_test.go

```diff
@@ -216,3 +252,87 @@ func TestPrebindPlugin(t *testing.T) {
 		cleanupPods(cs, t, []*v1.Pod{pod})
 	}
 }
+
+// TestUnreservePlugin tests invocation of un-reserve plugin
+func TestUnreservePlugin(t *testing.T) {
+	// TODO: register more plugin which would trigger un-reserve plugin
+	registry := framework.Registry{
+		unreservePluginName: NewUnreservePlugin,
+		prebindPluginName:   NewPrebindPlugin,
+	}
+
+	// Create the master and the scheduler with the test plugin set.
+	context := initTestSchedulerWithOptions(t,
+		initTestMaster(t, "unreserve-plugin", nil),
+		false, nil, registry, false, time.Second)
+	defer cleanupTest(t, context)
+
+	cs := context.clientSet
+	// Add a few nodes.
+	_, err := createNodes(cs, "test-node", nil, 2)
+	if err != nil {
+		t.Fatalf("Cannot create nodes: %v", err)
+	}
+
+	tests := []struct {
+		prebindFail   bool
+		prebindReject bool
+	}{
+		{
+			prebindFail:   false,
+			prebindReject: false,
+		},
+		{
+			prebindFail:   true,
+			prebindReject: false,
+		},
+		{
+			prebindFail:   false,
+			prebindReject: true,
+		},
+		{
+			prebindFail:   true,
+			prebindReject: true,
+		},
+	}
+
+	for i, test := range tests {
+		pbdPlugin.failPrebind = test.prebindFail
+		pbdPlugin.rejectPrebind = test.prebindReject
+
+		// Create a best effort pod.
+		pod, err := createPausePod(cs,
+			initPausePod(cs, &pausePodConfig{Name: "test-pod", Namespace: context.ns.Name}))
+		if err != nil {
+			t.Errorf("Error while creating a test pod: %v", err)
+		}
+
+		if test.prebindFail {
+			if err = wait.Poll(10*time.Millisecond, 30*time.Second, podSchedulingError(cs, pod.Namespace, pod.Name)); err != nil {
+				t.Errorf("test #%v: Expected a scheduling error, but didn't get it. error: %v", i, err)
+			}
+			if unresPlugin.numUnreserveCalled != pbdPlugin.numPrebindCalled {
```

Comment:
Is there any explanation why the unreserve plugin might be called more than once?

Reply:
From log, I find sometimes the test pod will be schedule twice before it be cleaned.
```
I0509 00:49:27.662644    7237 wrap.go:47] POST /api/v1/namespaces/test-2/pods: (1.030311ms) 201 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42654]
I0509 00:49:27.662832    7237 scheduling_queue.go:795] About to try and schedule pod test-2/test-pod
I0509 00:49:27.662850    7237 scheduler.go:452] Attempting to schedule pod: test-2/test-pod
I0509 00:49:27.662962    7237 scheduler_binder.go:256] AssumePodVolumes for pod "test-2/test-pod", node "test-node-0"
I0509 00:49:27.662980    7237 scheduler_binder.go:266] AssumePodVolumes for pod "test-2/test-pod", node "test-node-0": all PVCs bound and nothing to do
numPrebindCalled++ -> 1
I0509 00:49:27.663012    7237 framework.go:85] rejected by prebind-plugin at prebind: reject pod test-pod
E0509 00:49:27.663025    7237 factory.go:662] Error scheduling test-2/test-pod: rejected by prebind-plugin at prebind: reject pod test-pod; retrying
I0509 00:49:27.663043    7237 factory.go:720] Updating pod condition for test-2/test-pod to (PodScheduled==False, Reason=Unschedulable)
I0509 00:49:27.668967    7237 wrap.go:47] POST /api/v1/namespaces/test-2/events: (5.354584ms) 201 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42656]
I0509 00:49:27.669020    7237 wrap.go:47] GET /api/v1/namespaces/test-2/pods/test-pod: (5.473273ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42652]
I0509 00:49:27.669169    7237 wrap.go:47] PUT /api/v1/namespaces/test-2/pods/test-pod/status: (5.935527ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42654]
I0509 00:49:27.669322    7237 scheduling_queue.go:795] About to try and schedule pod test-2/test-pod
I0509 00:49:27.669347    7237 scheduler.go:452] Attempting to schedule pod: test-2/test-pod
numUnreserveCalled++ -> 1
I0509 00:49:27.669550    7237 scheduler_binder.go:256] AssumePodVolumes for pod "test-2/test-pod", node "test-node-0"
I0509 00:49:27.669570    7237 scheduler_binder.go:266] AssumePodVolumes for pod "test-2/test-pod", node "test-node-0": all PVCs bound and nothing to do
numPrebindCalled++ -> 2
I0509 00:49:27.669609    7237 framework.go:85] rejected by prebind-plugin at prebind: reject pod test-pod
E0509 00:49:27.669618    7237 factory.go:662] Error scheduling test-2/test-pod: rejected by prebind-plugin at prebind: reject pod test-pod; retrying
I0509 00:49:27.669638    7237 factory.go:720] Updating pod condition for test-2/test-pod to (PodScheduled==False, Reason=Unschedulable)
numUnreserveCalled++ -> 2
I0509 00:49:27.671425    7237 wrap.go:47] GET /api/v1/namespaces/test-2/pods/test-pod: (1.537755ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42652]
E0509 00:49:27.671603    7237 factory.go:686] pod is already present in unschedulableQ
I0509 00:49:27.671793    7237 wrap.go:47] PATCH /api/v1/namespaces/test-2/events/test-pod.159cf4497ef3b14b: (1.799554ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42656]
I0509 00:49:27.764167    7237 wrap.go:47] GET /api/v1/namespaces/test-2/pods/test-pod: (739.161µs) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42656]
reset numUnreserveCalled -> 0
reset numPrebindCalled -> 0
I0509 00:49:27.766347    7237 scheduling_queue.go:795] About to try and schedule pod test-2/test-pod
I0509 00:49:27.766376    7237 scheduler.go:448] Skip schedule deleting pod: test-2/test-pod
I0509 00:49:27.771321    7237 wrap.go:47] POST /api/v1/namespaces/test-2/events: (4.727587ms) 201 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42652]
I0509 00:49:27.771869    7237 wrap.go:47] DELETE /api/v1/namespaces/test-2/pods/test-pod: (7.36413ms) 200 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42656]
I0509 00:49:27.774353    7237 wrap.go:47] GET /api/v1/namespaces/test-2/pods/test-pod: (1.067354ms) 404 [scheduler.test/v0.0.0 (linux/amd64) kubernetes/$Format 127.0.0.1:42656]
danieldebug: pod is deleted
```

Reply:
@bsalamat Due to the time competition between :
    1. `scheduling_queue` retry to schedule the test pod
    2. `cleanupPods`

What we can do here is to check whether the times of `prebind` fails equal the times of `unreserve`. And I think it's safe and by design, right ?

Reply:
Correct. This is expected. Thanks for checking. More accurately, since the pod is rejected at pre-bind, the scheduler will retry scheduling it. The scheduler may retry the pod one or more times before we check the number of times `unreserve` is called. In order to make the test more robust, please change the condition to:

`if unresPlugin.numUnreserveCalled == 0 || unresPlugin.numUnreserveCalled != pbdPlugin.numPrebindCalled`

Reply:
Agree. Condition changed. PTAL